streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```
@@ -38,346 +40,644 @@
+            .collect(Collectors.toSet());
+    }
+
+    // 1. if there are any registered changelogs that needs initialization, try to initialize them first;
+    // 2. if all changelogs have finished, return early;
+    // 3. if there are any restoring changelogs, try to read from the restore consumer and process them.
+    public void restore() {
+        initializeChangelogs(registeredChangelogs());
+
+        if (allChangelogsCompleted()) {
+            log.info("Finished restoring all changelogs {}", changelogs.keySet());
             return;
         }
 
-        endOffsets.forEach((partition, endOffset) -> {
-            if (endOffset != null) {
-                final StateRestorer restorer = stateRestorers.get(partition);
-                final long offsetLimit = restorer.offsetLimit();
-                restoreToOffsets.put(partition, Math.min(endOffset, offsetLimit));
-            } else {
-                log.info("End offset cannot be found form the returned metadata; removing this partition from the current run loop");
-                initializable.remove(partition);
+        if (!restoringChangelogs().isEmpty()) {
+            final ConsumerRecords<byte[], byte[]> polledRecords = restoreConsumer.poll(pollTime);
+
+            for (final TopicPartition partition : polledRecords.partitions()) {
+                final ChangelogMetadata changelogMetadata = changelogs.get(partition);
+                if (changelogMetadata == null) {
+                    throw new IllegalStateException("The corresponding changelog restorer for " + partition +
+                        " does not exist, this should not happen.");
+                }
+                if (changelogMetadata.changelogState != ChangelogState.RESTORING) {
+                    throw new IllegalStateException("The corresponding changelog restorer for " + partition +
+                        " has already transited to completed state, this should not happen.");
```

Should we check and/or log the actual state rather than assume it would be `COMPLETED` (not `REGISTERED`)?

---

We are looping through `restoringChangelogs()` which should be those `RESTORING` ones only -- but note that even those paused changelogs are still in `RESTORING`.

---

Well if we assume these will all be `RESTORING` then why do the check at all -- if for some reason something went wrong and they are _not_ `RESTORING` as expected, it could be because the state was either `COMPLETED` or `REGISTERED`, but we don't know which

---

```suggestion
                        " has already transited to " + changelogMetadata.changelogState + " state, this should not happen.");
```

---

Yeah I think this check is more for discovering bugs and fail fast; I can update the error message as you suggested.